File: practice/22_Generate_Parentheses.kt
fun main() {
	println(generate(3))
}

fun generate(n: Int): List<String> {
	return helper(n, n, "")
}

fun helper(i: Int, j: Int, p: String):List<String> {
	var result = mutableListOf<String>()

	if(i==0 && j==0){
		return listOf(p)
	}

	if(i>0){
		result.addAll(helper(i-1, j, "$p("))
	}

	if(i<j &&j>0){
		result.addAll(helper(i, j-1, "$p)"))
	}

	return result
}

File: practice/22_Generate_Parentheses.js
function generate(n) {
    return helper(n, n, '');
}

function helper(i, j, p) {
    let result = [];
    if (i === 0 && j == 0) {
        return [p];
    }

    if (i > 0) {
        result = result.concat(helper(i - 1, j, p + '('));
    }
    if (j > 0 && i < j) {
        result = result.concat(helper(i, j - 1, p + ')'));
    }

    return result;
}

console.log(generate(3));
